velox/functions/lib/Re2Functions.cpp

```diff
@@ -666,6 +796,114 @@ FOLLY_ALWAYS_INLINE static bool isAsciiArg(
+  // The index in the input we are matching against.
+  auto currentMatchingOffsetInInput = 0;
+  auto searchLiteralPattern =
+      [&](const SubPatternMetadata& subPattern) -> int32_t {
+    auto index = inputView.find(
+        patternStrHeader + subPattern.start,
+        currentMatchingOffsetInInput,
+        subPattern.length);
+    if (index == std::string::npos) {
+      return -1;
+    }
+
+    return index;
+  };
+
+  // The index of the sub literal pattern we are matching.
+  auto currentPatternIndex =
+      (patternMetadata.subPatterns()[0].kind == kLiteralString ? 0 : 1);
+  auto currentSubPattern = patternMetadata.subPatterns()[currentPatternIndex];
+  auto end = input.size();
+  while (currentMatchingOffsetInInput < end) {
+    // Search current literal pattern to find out where to start matching
+    // the whole pattern.
+    currentMatchingOffsetInInput = searchLiteralPattern(currentSubPattern);
+    // If any literal pattern is not found, fails the matching.
+    if (currentMatchingOffsetInInput < 0) {
+      return false;
+    }
+
+    RelaxedPatternMatchResult matchResult =
+        RelaxedPatternMatchResult::failedForTooShortInput();
+    // Match the right part of the pattern(if there is a right part).
+    if (!isRightMostSubPattern(patternMetadata, currentPatternIndex)) {
+      matchResult = matchRelaxedFixedForward<isAscii>(
+          input,
+          patternMetadata,
+          currentMatchingOffsetInInput + currentSubPattern.length,
+          currentPatternIndex + 1);
+    }
+
+    // Match the left part of the pattern(if there is a left part).
+    // We only match the left part if:
+    // - Current pattern is the right most.
+    // - We successfully matched the right part of the pattern.
+    if (isRightMostSubPattern(patternMetadata, currentPatternIndex) ||
+        (matchResult.match() && currentPatternIndex > 0)) {
+      matchResult = matchRelaxedFixedBackward<isAscii>(
+          input,
+          patternMetadata,
+          currentMatchingOffsetInInput - 1,
+          currentPatternIndex - 1);
+    }
+
+    // Both the left and right part of the pattern match, the whole pattern
+    // matches.
+    if (matchResult.match()) {
+      return true;
+    }
+
+    if (matchResult.tooShortInputToMatch()) {
+      return false;
+    }
+
+    // Search the mismatch literal pattern to find the location where we can
+    // start matching.
+    currentPatternIndex = matchResult.failingPatternIndex();
+    currentSubPattern = patternMetadata.subPatterns()[currentPatternIndex];
+    // The mismatch pattern is not a literal means the input is not long
+    // enough.
+    if (currentSubPattern.kind != kLiteralString) {
+      return false;
+    }
+
+    // matchResult.cursor() is where we failed matching the sub-pattern, we
+    // start from `matchResult.cursor() + 1` to search for the sub-pattern.
+    currentMatchingOffsetInInput = matchResult.cursor() + 1;
```

Comment:
Do we always advance only by 1 byte? I thought we have to do that only when we failed to match left side, no? Also, in case of unicode, should we advance by 1 character?